shared/appengine.go

```diff
@@ -12,50 +12,46 @@ import (
 	"net/http"
 	"net/url"
 	"os"
+	"strings"
 	"time"
 
+	cloudtasks "cloud.google.com/go/cloudtasks/apiv2"
 	"github.com/google/go-github/v31/github"
 	"golang.org/x/oauth2"
 	apps "google.golang.org/api/appengine/v1"
+	taskspb "google.golang.org/genproto/googleapis/cloud/tasks/v2"
 
 	// TODO(#1747): Deprecate this library.
 	"google.golang.org/appengine/user"
 )
 
-// runtimeIdentity contains the identity of the current AppEngine service when
-// running on GAE, or empty when running locally.
-var runtimeIdentity struct {
-	LocationID  string
-	AppID       string
-	Service     string
-	Version     string
-	application *apps.Application
+type clientsImpl struct {
+	cloudtasks *cloudtasks.Client
 }
 
-func init() {
-	// Env vars available on GAE:
-	// https://cloud.google.com/appengine/docs/standard/go/runtime#environment_variables
-	// Note: the "region code" part of GAE_APPLICATION is NOT location ID.
-	if proj := os.Getenv("GOOGLE_CLOUD_PROJECT"); proj != "" {
-		runtimeIdentity.AppID = proj
-		runtimeIdentity.Service = os.Getenv("GAE_SERVICE")
-		if runtimeIdentity.Service == "" {
-			panic("Missing environment variable: GAE_SERVICE")
-		}
-		runtimeIdentity.Version = os.Getenv("GAE_VERSION")
-		if runtimeIdentity.Version == "" {
-			panic("Missing environment variable: GAE_VERSION")
-		}
-		service, err := apps.NewService(context.Background())
-		if err != nil {
-			panic(err)
-		}
-		runtimeIdentity.application, err = service.Apps.Get(proj).Do()
-		if err != nil {
-			panic(err)
-		}
-		runtimeIdentity.LocationID = runtimeIdentity.application.LocationId
+// Clients is a singleton containing heavyweight (e.g. with connection pools)
+// clients that should be bound to the runtime instead of each request in order
+// to be reused. They are initialized and authenticated at startup using the
+// background context; each request should use its own context.
+var Clients clientsImpl
+
+// Init initializes all clients in Clients. If an error is encountered, it
+// returns immediately without trying to initialize the remaining clients.
+func (c *clientsImpl) Init(ctx context.Context) (err error) {
+	if runtimeIdentity.AppID == "" {
+		// When running in dev_appserver, do not create real clients.
+		return nil
 	}
+	c.cloudtasks, err = cloudtasks.NewClient(ctx)
+	return err
+}
+
+// Close closes all clients in Clients. It must be called once and only once
+// before the server exits. Do not use AppEngineAPI afterwards.
+func (c *clientsImpl) Close() error {
+	err := c.cloudtasks.Close()
```

Comment:
when `runtimeIdentity.AppID == ""` and return `nil`, `defer shared.Clients.Close()` will still be called. c.cloudtasks.Close() might cause a nil pointer exception

Reply:
Aha great find! Could you send a PR to fix it up?